dali/pipeline/operators/decoder/cache/decoder_cache_factory.h

```diff
@@ -0,0 +1,55 @@
+// Copyright (c) 2019, NVIDIA CORPORATION. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef DALI_PIPELINE_OPERATORS_DECODER_CACHE_DECODER_CACHE_FACTORY_H_
+#define DALI_PIPELINE_OPERATORS_DECODER_CACHE_DECODER_CACHE_FACTORY_H_
+
+#include <memory>
+#include <string>
+#include <map>
+#include <mutex>
+#include "dali/pipeline/operators/decoder/cache/decoder_cache.h"
+
+namespace dali {
+
+class DLL_PUBLIC DecoderCacheFactory {
+ public:
+  DLL_PUBLIC static inline DecoderCacheFactory& Instance(int device_id) {
+    static std::mutex __mutex;
+    using DeviceId = int;
+    static std::map<DeviceId, DecoderCacheFactory> __cache_factory_map;
+    std::unique_lock<std::mutex> lock(__mutex);
+    return __cache_factory_map[device_id];
+  }
+
+  DLL_PUBLIC std::shared_ptr<DecoderCache> Init(
```

Comment:
The whole flow with `IsInitialized` -> Get/Init creates a race condition - whoever calls `Init` first will set the cache parameters. What happens if there are two decoders in the pipeline? I think the cache should have larges requested size and lowest requested threshold.

Reply:
I don't think it's good that we play with reallocating the buffer if another instance wants a bigger one. Not really sure how to deal with this case though

Reply:
done